server/internal/mcp/rpc_resources_read.go

```diff
@@ -0,0 +1,255 @@
+package mcp
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v5/pgxpool"
+
+	"github.com/speakeasy-api/gram/server/internal/billing"
+	"github.com/speakeasy-api/gram/server/internal/conv"
+	"github.com/speakeasy-api/gram/server/internal/gateway"
+	"github.com/speakeasy-api/gram/server/internal/mv"
+	"github.com/speakeasy-api/gram/server/internal/o11y"
+	"github.com/speakeasy-api/gram/server/internal/oops"
+	"github.com/speakeasy-api/gram/server/internal/toolsets"
+	"github.com/speakeasy-api/gram/server/internal/urn"
+)
+
+type resourceReadParams struct {
+	URI string `json:"uri"`
+}
+
+type resourceReadResult struct {
+	Contents []resourceContent `json:"contents"`
+}
+
+type resourceContent struct {
+	URI      string  `json:"uri"`
+	Name     *string `json:"name,omitempty"`
+	MimeType *string `json:"mimeType,omitempty"`
+	Text     string  `json:"text,omitempty"`
+	Blob     string  `json:"blob,omitempty"`
+}
+
+func handleResourcesRead(ctx context.Context, logger *slog.Logger, db *pgxpool.Pool, payload *mcpInputs, req *rawRequest, toolProxy *gateway.ToolProxy, env gateway.EnvironmentLoader, billingTracker billing.Tracker, billingRepository billing.Repository) (json.RawMessage, error) {
```

Comment:
Given resources are ONLY served from MCP. I could see us adding more MCP resource specific utility here. This is where we are doing the binary type detection right now.